src/plugins/visualizations/public/legacy/build_pipeline.ts

```diff
@@ -550,3 +577,93 @@ export const buildPipeline = async (
 
   return pipeline;
 };
+
+export async function buildVisExpression(
+  vis: Vis,
+  params: {
+    timefilter: TimefilterContract;
+    timeRange?: TimeRange;
+    abortSignal?: AbortSignal;
+  }
+) {
+  const { indexPattern, savedSearchId, searchSource } = vis.data;
+  const { uiState } = vis;
+  const query = searchSource!.getField('query');
+  const filters = searchSource!.getField('filter');
+
+  // context
+  const pipeline = buildExpression([buildExpressionFunction('kibana', {})]);
+  const kibanaContext = buildExpressionFunction('kibana_context', {});
+
+  if (query && typeof query === 'string') {
+    // TODO:
+    // The typings for `Query` are `string | { [key: string]: any }
+    // Will it ever be in object format when coming from vis.data?
+    kibanaContext.addArgument('q', query);
+  }
+
+  if (filters) {
+    kibanaContext.addArgument('filters', JSON.stringify(filters));
+  }
+
+  if (savedSearchId) {
+    kibanaContext.addArgument('savedSearchId', savedSearchId);
+  }
+
+  // request handler
+  if (vis.type.requestHandler === 'courier' && indexPattern?.id) {
+    pipeline.functions.push(
+      buildExpressionFunction('esaggs', {
+        index: indexPattern.id,
+        metricsAtAllLevels: vis.isHierarchical(),
+        partialRows: vis.type.requiresPartialRows || vis.params.showPartialRows || false,
+        aggConfigs: JSON.stringify(vis.data.aggs?.aggs || '[]'),
+      })
+    );
+  }
+
+  const schemas = getSchemas(vis, {
+    timeRange: params.timeRange,
+    timefilter: params.timefilter,
+  });
+
+  if (buildPipelineVisFunction[vis.type.name]) {
+    // TODO: temporarily parsing string here
+    const { chain } = parseExpression(
+      buildPipelineVisFunction[vis.type.name](vis.params, schemas, uiState)
+    );
+    chain.forEach(({ function: fnName, arguments: args }) =>
+      pipeline.functions.push(buildExpressionFunction(fnName, args))
+    );
+  } else if (vislibCharts.includes(vis.type.name)) {
+    const visConfig = { ...vis.params };
+    visConfig.dimensions = await buildVislibDimensions(vis, params);
+
+    pipeline.functions.push(
+      buildExpressionFunction('vislib', {
+        type: vis.type.name,
+        visConfig: JSON.stringify(visConfig),
+      })
+    );
+  } else if (vis.type.toExpression) {
+    pipeline.functions.push(await vis.type.toExpression(vis, params));
```

Comment:
we might want to return expression from toExpression, so chart can optionally return more than one function, so we should concat the chains